### PR TITLE
Implement CLI interface for pose2d estimation

### PR DESCRIPTION
**Pull request by julien-h**
Implements a CLI interface using `argparse` in `./CLI/main.py`.

**Reference:** I took the code in `GUI/main.py/DrosophAnnot.pose2d_estimation` to find which arguments are parsed and which are set in code before calling `pose2d.drosophilia.main`.

**Example usage:** 
The following should achieve the same result (no `diff` output between generated `.pkl` files):

```
> deepfly ./data/test 15
then click on pos2d estimate
```

and

```
python -m deepfly.CLI.main ./data/test -n 15 --resume ./weights/sh8_deepfly.tar
```

**Differences**

- The default value for the argument `--resume` is not the same in Config.py and in ArgParse.py hence the need to specify it to obtain the same output.
- We can make the max number of images (-n 15) a positional argument, but in this case it will be required. We can take a convention such as "inf" or "0" to mean no maximum.

**Side remark**

Note that (as in the GUI code) the arguments --checkpoint and --unlabeled are overriden after argparse parses them.

### CODE REVIEW

**Comment by julien-h** [comment]
description updated.

**Comment by semihgunel** [comment]
Cool work. We should also consider decreasing the number of necessary arguments.
1. We can eliminate `--resume` by setting it the default value on `ArgParse.py` from `Config.py`.
2. My understanding is `--n` is not necessary at all, as it would process all the data because the default is `inf`. Is this correct?

A nice addition would be enabling recursive processing so that it would automatically process every folder inside the tree (perhaps which has some images inside?). We can have a `--recursive` flag to enable this.

**Comment by julien-h** [comment]
Yes by convention arguments that start with a dash `-` are optional. The only necessary argument is thus the input folder:
```
python -m deepfly.CLI.main ./data/test
```

- The default value for `-n` is indeed `inf` so that we process all available images.
- `--resume` is not required either. If we don't add it I think the program just starts from scratch. I will edit the `ArgParse.py` file to fetch its default value from `Config.py`.

And will look into the recursive stuff you suggested.

File: deepfly/CLI/main.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        "input_folder", 
+        help="Folder containing unlabeled images."
+    )
+    parser.add_argument(
+        "-n", "--num-images-max",
+        help="Maximal number of images to process.",
+        default=math.inf,
+        type=int
+    )
+    parser.add_argument( # TODO: there is "--stacks" in ArgParse, is it the same?
```
Comment (semihgunel) [inline deepfly/CLI/main.py]: It is the same thing with `--stacks`, we should use that one instead. you can remove this argument.

**Comment by julien-h** [comment]
Just to be ultra clear about the argument `-n 15`.

- If we want it optional, the convention is to keep it like this `-n 15`. Minimal command call is thus:
```
cli.py folder
```
The default is to process all images and we can specify a max number with `-n 15` for instance.

- If we make it compulsory, we can make it a positional argument and remove the dash and write only `15` instead of `-n 15`. Minimal command call becomes:

```
cli.py folder inf  // for no maximal number
cli.py folder 15  // for 15 images max
```

**Comment by semihgunel** [comment]
I think we shouldn't make things compulsory as this type of software is used by non-technical people in general. I think the current version is optimal. 

We can merge the current version with:
1. recursive handling with a `--recursive` flag,
2. setting the default value of `--resume`, to make sure CLI would work even without setting it,
3. and removing the unnecessary argument.

**Comment by julien-h** [comment]
There is a recursive option in `ArgParse.py line 164`. Below is summary of what the code does when set. Tell me if you want to change anything.

```python
parser.add_argument(
    "--unlabeled-recursive",
    dest="unlabeled_recursive",
    action="store_true",
    default=False,
)
```

When the option is set, it is turned into a list of folders in `drosophilia.py line 153`:

```python
if args.unlabeled_recursive:
    print("Recursive search in unlabeled folder")
    unlabeled_folder_list = find_leaf_recursive(args.unlabeled)
    unlabeled_folder_list = [
        path for path in unlabeled_folder_list if "images" in path
    ]
else:
    unlabeled_folder_list = [args.unlabeled]
```

Then the program will iterate over the list and basically run a copy of itself in each folder. The general outline is:

```python
for unlabeled_folder in unlabeled_folder_list:
    unlabeled_loader = DataLoader(...)
    loss,acc,pred = validate(..., save_path=unlabeled_folder)
    save_dict(args.data_folder"/"unlabeled_folder"/preds_{}.pkl")
```

The behavior of `max_num_images` seem undefined in this case and my guess is that the program will process at most `max_num_images` in *each* subfolder.

**Comment by semihgunel** [comment]
with c073d31 I changed the program name to df3d, and added df3d-cli.

**Comment by semihgunel** [comment]
> There is a recursive option in `ArgParse.py line 164`. Below is summary of what the code does when set. Tell me if you want to change anything.
> 
> ```python
> parser.add_argument(
>     "--unlabeled-recursive",
>     dest="unlabeled_recursive",
>     action="store_true",
>     default=False,
> )
> ```
> 
> When the option is set, it is turned into a list of folders in `drosophilia.py line 153`:
> 
> ```python
> if args.unlabeled_recursive:
>     print("Recursive search in unlabeled folder")
>     unlabeled_folder_list = find_leaf_recursive(args.unlabeled)
>     unlabeled_folder_list = [
>         path for path in unlabeled_folder_list if "images" in path
>     ]
> else:
>     unlabeled_folder_list = [args.unlabeled]
> ```
> 
> Then the program will iterate over the list and basically run a copy of itself in each folder. The general outline is:
> 
> ```python
> for unlabeled_folder in unlabeled_folder_list:
>     unlabeled_loader = DataLoader(...)
>     loss,acc,pred = validate(..., save_path=unlabeled_folder)
>     save_dict(args.data_folder"/"unlabeled_folder"/preds_{}.pkl")
> ```
> 
> The behavior of `max_num_images` seem undefined in this case and my guess is that the program will process at most `max_num_images` in _each_ subfolder.

Yes, I remember implementing this. This is rather a poor implementation as it merely finds leafs in the filesystem tree and processes those. It is well possible images are in non-leaf folders. Yet, I think this is enough for now. Can you still add the `--recursive` and assign it to `--unlabeled-recursive`?

**Comment by julien-h** [comment]
> Can you still add the --recursive and assign it to --unlabeled-recursive?

Yes. And I can re-implement the recursive behavior from scratch, parsing all images in the folder and its subdirectories if you want.

**Comment by semihgunel** [comment]
> > Can you still add the --recursive and assign it to --unlabeled-recursive?
> 
> Yes. And I can re-implement the recursive behavior from scratch, parsing all images in the folder and its subdirectories if you want.

Lets pass this for now unless you specifically want to implement this :)

**Comment by julien-h** [comment]
I will do some additional tests to make sure GUI output and CLI output are the same before merging

**Comment by semihgunel** [comment]
I added a single *.pkl file with 645ee91. It needs to stay as it is used for template for post-processing (although conflicting with your current .gitignore).

**Comment by julien-h** [comment]
Oh sorry, I did it automatically out of habit (no datafile in git repo).
Shall I undo my changes to `.gitignore` ?

**Comment by semihgunel** [comment]
> Oh sorry, I did it automatically out of habit (no datafile in git repo).
> Shall I undo my changes to `.gitignore` ?

I think that file should be the exception. So we can keep .gitignore. (I don't think .gitignore will complain about the files already in the repo).

**Comment by julien-h** [comment]
Ok, I ran the code locally and everything seems fine. I'll come at the lab when I can to compare with the GUI's output.

**Comment by julien-h** [comment]
Okay, I came at the lab and ran the test. All the arguments are the same now that I have set the default value of checkpoint to false.

We can merge. Do you want me to rebase and clean the history before?

**Pull request merged**